### ISSUE TITLE
[question] How to specify settings for package dependencies?

### ISSUE THREAD

**Issue by freihamer** [conan-io/conan#14212]
### What is your question?

Hi,

I have a number of baremetal, embedded software applications, which I need to build for different hardware platforms. I am creating conan packages from each of these, using custom settings I introduced in settings_user.yml

```
os:
    baremetal:
        platform:
            E0055:
                revision: [null, "01.01", "02.02"]
            E0064:
                revision: [null, "01.01", "02.01"]
            E0065:
                revision: [null, "02.02", "03.01"]
            null:
```

Then, I want to have a "parent" package which bundles all my "child" packages. The recipe for this one is very simple at the moment:

```
class ParentPkgRecipe(ConanFile):
    name = "parent-pkg"
    version = "1.0"

    def requirements(self):
        self.requires("child-pkg-1/1.1.0")
        self.requires("child-pkg-2/1.0.4")
        self.requires("child-pkg-3/1.2.2")
```
How can I create my parent package in a way that for example:

- For `child-pkg-1/1.1.0` the build for `os.platform=E0055`, `os.platform.revision=01.01` is used
- For `child-pkg-2/1.0.4` the build for `os.platform=E0064`, `os.platform.revision=02.01` is used
- For` child-pkg-3/1.2.2` the build for `os.platform=E0065`, `os.platform.revision=03.01` is used

In other words, use specific settings to include each child. Not just apply settings to the parent and the same will be used for all childs. Actually my custom settings are not even applicable to the parent package, as it is just a bundle.

Thanks,
Luca

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by czoido** [comment]
Hi @lucabarbiero,

Thanks for the question. You can use the syntax `<pattern>:<setting>=<value>` in your case `child-pkg-2/1.0.4:os.platform=E0064`

For example in this case, if you pass that in the command line (you can also have this in profiles) and use a `*` as version to affect to all versions:

```
conan install . --build=missing -s="os=baremetal" -s="child-pkg-1/*:os.platform=E0055" -s="child-pkg-1/*:os.platform.revision=01.01" -s="child-pkg-2/*:os.platform=E0064" -s="child-pkg-2/*:os.platform.revision=02.01" -s="child-pkg-3/*:os.platform=E0065" -s="child-pkg-3/*:os.platform.revision=03.01"
```

Should make it. Hope this helps!

**Comment by freihamer** [comment]
Hi @czoido,

Great thanks, that is what I needed!

Now I need my parent package to run some script on the binaries which are part of the required child packages, then package the outputs in the parent package. How do I do this?

I am trying to export my script `do_some_stuff.bat` from the recipe folder to the cache, but this is not happening when I run conan create... why? I get error when attempting to run that script in the build() and I after checking the cache I can confirm that the script is not there where it should.

If I just run conan build instead, the script is of course found because it is just in the recipe folder, but where are the binaries from the required child packages? Say for example that `binary-file-1.hex` is in `child-pkg-1/1.1.0` and so on.

Can you correct the mistakes, nonsenses and missing parts in the following recipe, so that conan build and conan create will work?
```
class ParentPkgRecipe(ConanFile):
    name = "parent-pkg"
    version = "1.0"
    
    exports = "do_some_stuff.bat"

    def requirements(self):
        self.requires("child-pkg-1/1.1.0")
        self.requires("child-pkg-2/1.0.4")
        self.requires("child-pkg-3/1.2.2")
    
    def build(self):
        self.run("do_some_stuff.bat -i binary-file-1.hex -o binary-file-1-mod.hex")
        self.run("do_some_stuff.bat -i binary-file-2.hex -o binary-file-2-mod.hex")
        self.run("do_some_stuff.bat -i binary-file-3.hex -o binary-file-3-mod.hex")

    def package(self):
        copy(self, "binary-file-1-mod.hex", self.build_folder, self.package_folder)
        copy(self, "binary-file-2-mod.hex", self.build_folder, self.package_folder)
        copy(self, "binary-file-3-mod.hex", self.build_folder, self.package_folder)
```

**Comment by memsharded** [comment]
Hi @lucabarbiero 

The ``exports`` is for things that the "recipe" itself needs, like the Python code. Things like ``conandata.yml``.
But the ``.bat`` is more like a source script. It doesn't belong to the recipe it belongs to the "contents" of the recipe. So this can be fixed by using ``exports_sources = `` instead.


Besides that, a couple of hints:
- the ``binary-file-x.hex`` and other files will not be in the path of the current build, unless you copy it first
- You can probably do that ``copy()`` in the ``generate()`` method, so it should work both in the local build and in ``create`` (I guess that you didn't get to this point yet)
- As the parent package seems to be re-packaging the changed contents, in Conan 2.0 you might be able to optimize the need for the transitive dependencies binaries using the ``traits`` (like ``libs=False, headers=False, transitive_libs=False, ...``

**Comment by freihamer** [comment]
Hi @memsharded,

Thanks I had read about `exports ` vs `exports_sources` but I managed to forget the difference soon after.

About the second part I solved with a different approach, that is instead of touching on the generate() and build() methods, I created a custom deployer, which means that my modified binaries will not be stored with the package but will be created at deployment time, which works and it even makes more sense if I think about it. The example provided [here](https://docs.conan.io/2/examples/extensions/deployers/sources/custom_deployer_sources.html) helped me to understand how to get artifacts from dependencies and do something on them before copying to the deployment output folder.

**Comment by memsharded** [comment]
Good, thanks for the feedback!
